src/esp/assets/BaseMesh.cpp

```diff
@@ -19,5 +24,99 @@ bool BaseMesh::setMeshType(SupportedMeshType type) {
   return true;
 }
 
+void BaseMesh::buildMeshColors(
+    const Mn::Trade::MeshData& srcMeshData,
+    bool convertToSRGB,
+    Cr::Containers::Array<Mn::Color3ub>& meshColors) const {
+  /* Assuming colors are 8-bit RGB to avoid expanding them to float and then
+     packing back */
+
+  if (convertToSRGB) {
+    auto colors = srcMeshData.colorsAsArray();
+    for (std::size_t i = 0; i != colors.size(); ++i) {
+      meshColors[i] = colors[i].rgb().toSrgb<Mn::UnsignedByte>();
+    }
+  } else {
+    Mn::Math::packInto(Cr::Containers::arrayCast<2, float>(
+                           stridedArrayView(srcMeshData.colorsAsArray()))
+                           .except({0, 1}),
+                       Cr::Containers::arrayCast<2, Mn::UnsignedByte>(
+                           stridedArrayView(meshColors)));
+  }
+}  // BaseMesh::buildMeshColors
+
+void BaseMesh::buildSemanticOBBs(
+    const std::vector<vec3f>& vertices,
+    const std::vector<uint16_t>& vertSemanticIDs,
+    const std::vector<std::shared_ptr<esp::scene::SemanticObject>>& ssdObjs,
+    const std::string& msgPrefix) const {
+  // build per-SSD object vector of known semantic IDs
+  std::size_t numSSDObjs = ssdObjs.size();
+  // no semantic ID 0
+  std::vector<int> semanticIDToSSOBJidx(numSSDObjs + 1);
+  for (int i = 0; i < numSSDObjs; ++i) {
+    const auto& ssdObj = *ssdObjs[i];
+    int semanticID = ssdObj.semanticID();
+    if (semanticIDToSSOBJidx.size() <= semanticID) {
+      semanticIDToSSOBJidx.resize(semanticID + 1);
+    }
+    semanticIDToSSOBJidx[semanticID] = i;
+  }
+
+  // aggegates of per-semantic ID mins and maxes
+  std::vector<esp::vec3f> vertMax(
+      semanticIDToSSOBJidx.size(),
+      {-Mn::Constants::inf(), -Mn::Constants::inf(), -Mn::Constants::inf()});
+  std::vector<esp::vec3f> vertMin(
+      semanticIDToSSOBJidx.size(),
+      {Mn::Constants::inf(), Mn::Constants::inf(), Mn::Constants::inf()});
+  std::vector<int> vertCounts(semanticIDToSSOBJidx.size());
+
+  // for each vertex, map vert min and max for each known semantic ID
+  // Known semantic IDs are expected to be contiguous, and correspond to the
+  // number of unique ssdObjs mappings.
+
+  for (int vertIdx = 0; vertIdx < vertSemanticIDs.size(); ++vertIdx) {
+    // semantic ID on vertex - valid values are 1->semanticIDToSSOBJidx.size().
+    // Invalid/unknown semantic ids are > semanticIDToSSOBJidx.size()
+    const auto semanticID = vertSemanticIDs[vertIdx];
+    if ((semanticID > 0) && (semanticID <= semanticIDToSSOBJidx.size())) {
+      const auto vert = vertices[vertIdx];
+      // FOR VERT-BASED OBB CALC
+      // only support bbs for known colors that map to semantic objects
+      vertMax[semanticID] = vertMax[semanticID].cwiseMax(vert);
+      vertMin[semanticID] = vertMin[semanticID].cwiseMin(vert);
+      vertCounts[semanticID] += 1;
+    }
```

Comment:
No logging for invalid semantics?

Reply:
Not there, since it could be potentially thousands of vertices with unknown/inappropriate semantic IDs.   The caller of this function is a more appropriate location for this kind of logging, though, where I have a record of unknown colors found on vertices.  

This kind of reporting is in progress, currently.  Both colors found on verts not found in Semantic scene lexicon, and colors in lexicon that were not found in the mesh will be listed. Ideally the user would be given the option of generating the report and saving it to disk so all the extra console clutter can be avoided.  I'm working on this.